perma_web/api/authorizations.py

```diff
@@ -1,124 +1,107 @@
-from perma.models import Link, Folder, VestingOrg
+from perma.models import Link, Folder
 from tastypie.authorization import ReadOnlyAuthorization
-from tastypie.exceptions import Unauthorized
+from tastypie.exceptions import Unauthorized, BadRequest
 
 
-class DefaultAuthorization(ReadOnlyAuthorization):
+class FolderAuthorization(ReadOnlyAuthorization):
 
-    def create_detail(self, object_list, bundle):
-        if not bundle.request.user.is_authenticated():
-            raise Unauthorized("You must be a registered user.")
-
-        return True
-
-    def update_detail(self, object_list, bundle):
-        if bundle.obj.created_by != bundle.request.user:
-            raise Unauthorized("Sorry, you don't have permission")
-
-        return True
+    def can_access(self, user, obj):
+        # staff can access any folder
+        if user.is_staff:
+            return True
 
-    delete_detail = update_detail
+        # private folders
+        if obj.owned_by_id == user.pk:
+            return True
 
+        # shared folders
+        elif obj.vesting_org_id:
+            if user.is_vesting_org_member():
+                return user.vesting_org_id == obj.vesting_org_id
+            elif user.is_registrar_member():
+                return user.registrar_id == obj.vesting_org.registrar_id
 
-class FolderAuthorization(DefaultAuthorization):
+        return False
 
-    def can_access(self, user, obj):
-        try:
-            # returns true if exists
-            return bool(Folder.objects.get(Folder.objects.user_access_filter(user),
-                                           pk=obj.pk))
-        except Folder.DoesNotExist:
-            return False
+    def read_list(self, object_list, bundle):
+        # /folders endpoint shouldn't list anything, but /folders/<id>/folders should
+        if not bundle.parent_object:
+            raise Unauthorized()
+        return object_list
 
     # NOTE - this is called by obj_update and obj_delete before it uses their auth methods
     # ex: https://github.com/toastdriven/django-tastypie/blob/master/tastypie/resources.py#L2203
     def read_detail(self, object_list, bundle):
         if not self.can_access(bundle.request.user, bundle.obj):
-            raise Unauthorized("Sorry, you don't have access to that folder.")
+            raise Unauthorized()
 
         return True
 
     def delete_detail(self, object_list, bundle):
+
+        if not self.can_access(bundle.request.user, bundle.obj):
+            raise Unauthorized()
+
         if bundle.obj.is_shared_folder:
-            raise Unauthorized("Shared folders cannot be deleted.")
+            raise BadRequest("Shared folders cannot be deleted.")
```

Comment:
Seems weird to throw `BadRequest` in an `Authorization` rather than a `Validation`. I seem to remember talking about this, and if there was a reason such as an early exit to avoid an expensive DB call that would take place before reaching the `Validation`, then I'm okay with that. Otherwise, it should probably be moved to a `Validation`.